runners/spark/src/main/java/org/apache/beam/runners/spark/translation/EvaluationContext.java

```diff
@@ -248,21 +242,40 @@ public State getState() {
 
   @Override
   public State cancel() throws IOException {
-    throw new UnsupportedOperationException(
-        "Spark runner EvaluationContext does not support cancel.");
+    return cancel(true);
+  }
+
+  private State cancel(boolean gracefully) throws IOException {
+    if (isStreamingPipeline()) {
+      if (!state.isTerminal()) {
+        jssc.stop(false, gracefully);
+        state = State.CANCELLED;
+      }
+      return state;
+    } else {
+      throw new UnsupportedOperationException(
+              "Spark runner EvaluationContext does not support cancel.");
```

Comment:
This is true for batch, and probably simply a stub.
Since batch execution is blocking, you should never get there..